platform/ext/target/nordic_nrf/common/core/tfm_hal_its_encryption.c

```diff
@@ -0,0 +1,296 @@
+/*
+ * Copyright (c) 2023 Nordic Semiconductor ASA.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "platform/include/tfm_platform_system.h"
+#include "platform/include/tfm_hal_its.h"
+#include "psa/crypto.h"
+
+#include "nrf_cc3xx_platform_derived_key.h"
+#include "nrf_cc3xx_platform.h"
+#include "nrf_cc3xx_platform_kmu.h"
+#include <string.h>
+
+#define ITS_ENCRYPTION_SUCCESS 0
+
+#define HUK_KMU_SLOT 2
+#define HUK_KMU_SIZE_BITS 128
+
+/* Global encryption counter which resets per boot. The counter ensures that
+ * the nonce will not be identical for consecutive file writes during the same
+ * boot.
+ */
+static uint32_t g_enc_counter;
+
+/* The global nonce seed which is fetched once in every boot. The seed is used
+ * as part of the nonce and allows the platforms to diversify their nonces
+ * across resets. Note that the way that this seed is generated is platform
+ * specific, so the diversification is optional.
+ */
+static uint8_t g_enc_nonce_seed[TFM_ITS_ENC_NONCE_LENGTH -
+                                sizeof(g_enc_counter)];
+
+/* As TFM_ITS_ENC_NONCE_LENGTH is configurable but this implementation expects
+ * the seed to be 8 bytes and the nonce length to always be 12.
+ */
+#if TFM_ITS_ENC_NONCE_LENGTH != 12
+#error "This implementation only supports a ITS nonce of size 12"
+#endif
+
+
+enum tfm_hal_status_t tfm_hal_its_aead_generate_nonce(uint8_t *nonce,
+                                                      size_t nonce_size)
+{
+    int err;
+
+    if(nonce == NULL){
+        return TFM_HAL_ERROR_INVALID_INPUT;
+    }
+
+    if(nonce_size < sizeof(g_enc_nonce_seed) + sizeof(g_enc_counter)){
+        return TFM_HAL_ERROR_INVALID_INPUT;
+    }
+
+    if (g_enc_counter == 0) {
+        if (sizeof(g_enc_nonce_seed) != 8) {
+            return TFM_HAL_ERROR_GENERIC;
+        }
+        /* The cc3xx driver is used directly as the PSA APIs depends on ITS, so
+         * using PSA APIs would result in circular dependencies */
+        err = nrf_cc3xx_platform_get_nonce_seed(g_enc_nonce_seed);
+        if (err != 0) {
+            return TFM_HAL_ERROR_GENERIC;
+        }
+    }
+
+    memcpy(nonce, g_enc_nonce_seed, sizeof(g_enc_nonce_seed));
+    memcpy(nonce + sizeof(g_enc_nonce_seed),
+               &g_enc_counter,
+               sizeof(g_enc_counter));
+
```

Comment:
```
/* To avoid wrap-around of the g_enc_counter and subsequent re-use of the nonce
 * we check for a counter value so large that it could only be nefariously produced.
 * /
if(g_enc_counter > 100 * 1000 * 1000) {
   return TFM_HAL_ERROR_GENERIC;
}
```

An attacker can construct input such that tfm_hal_its_aead_generate_nonce succeeds, but a later stage to fails, and then increment the counter until we re-use the nonce.

Reply:
I suppose this could should be higher up, as a part of input validation.